### DISCUSSION TITLE
Tooltip onOpenChange is not syncing the `open` value correctly

### DISCUSSION THREAD

**Discussion by p10ns11y** [radix-ui/primitives#2707] [Help] [answered]
It make sense that `onOpenChange` is not called during initial render
Can you clarify if I am right that I expect the `onOpenChange` should be called with the current value of `open` during state update 

Here is the example [codesandbox](https://codesandbox.io/p/devbox/vibrant-frog-h69swq?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clsn2al6300073b6kb1zerprn%2522%252C%2522sizes%2522%253A%255B57.86549708%252C42.13450292%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clsn2al6300023b6kaauld59k%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clsn2al6300043b6kndhcbk2p%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clsn2al6300063b6keyzu0s3t%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B56.29522939647298%252C43.70477060352702%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clsn2al6300023b6kaauld59k%2522%253A%257B%2522id%2522%253A%2522clsn2al6300023b6kaauld59k%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsn2hzng009y3b6kce8o65un%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A15%252C%2522startColumn%2522%253A25%252C%2522endLineNumber%2522%253A15%252C%2522endColumn%2522%253A25%257D%255D%252C%2522filepath%2522%253A%2522%252FApp.jsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clsn2hzng009y3b6kce8o65un%2522%257D%252C%2522clsn2al6300063b6keyzu0s3t%2522%253A%257B%2522id%2522%253A%2522clsn2al6300063b6keyzu0s3t%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsn2al6300053b6k0ul1mkdi%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clsn2al6300053b6k0ul1mkdi%2522%257D%252C%2522clsn2al6300043b6kndhcbk2p%2522%253A%257B%2522id%2522%253A%2522clsn2al6300043b6kndhcbk2p%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsn2al6300033b6kgf0mcoej%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clsn2al6300033b6kgf0mcoej%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D).

In the example, I am trying to implement tooltip with `click` event without two much tweaks and I expect the state `open`'s value is correct in the `onOpenChange` callback and it is not. I suspect other default pointer events are somehow blocking or causing state syncing broken.

- `onOpenChange` is called with `true` on initial hover (should all pointer events prevented?)
- Further clicks toggles the tooltip as expected (toggles between open and close) (visibly works)
- `onOpenChange` never called with `true` anymore even when the state `open` become `true` (internal state sync is still an issue)

Also wondering, what is the use case of `onOpenChange`, ideally one could sync  react state `open` in `useEffect`.  Does `onOpenChange` to make sure not letting the developer to manage such side effects and rather keep it predictable callback

**Comment by benoitgrelard** [comment]
That sandbox doesn't work so I'm not able to see what you mean.

**Reply by benoitgrelard** [reply]
This is probably related to what you are asking, see my answer there it might explain things: https://github.com/radix-ui/primitives/discussions/2195#discussioncomment-6126147

**Comment by p10ns11y** [comment]
Sorry about the link @benoitgrelard , it was private before and now it can be accessed.
I have previously checked the comment https://github.com/radix-ui/primitives/discussions/2195#discussioncomment-6126147.
The issue I am reporting is different here and it is not about `onOpenChange` is not being called initially.
It is about the value of `open` is not reflecting tooltip's actual visibility when using controlled props.

**Reply by benoitgrelard** [reply]
Why are you trying to open the tooltip on click rather than its normal behaviour?

**Reply by p10ns11y** [reply]
@benoitgrelard Use case is to use it in mobile devices (I am aware that it is better to avoid tooltips in mobile devices). Feel free to suggest any other alternative UI/UX solutions

**Reply by p10ns11y** [reply]
Used ChatGPT to figured out some alternatives

`Is Toggle section is good alternative?`

**ChatGPT**
Yes, toggle sections can be a good alternative to tooltips in certain contexts, especially on mobile or touch devices. Here's why:

**Visibility and Accessibility:** Toggle sections provide a more visible and accessible way to present additional information compared to tooltips, which often require a hover action. Users can easily see and interact with toggle sections, making them more intuitive, especially on touch-based devices.

**Content Presentation**: Toggle sections allow for the presentation of more extensive content or details in a structured and organized manner. Users can expand or collapse the section to access the information they need, providing a more comprehensive understanding of the content.

**Consistency and Predictability**: Toggle sections typically behave consistently across different devices and platforms, ensuring a predictable user experience. Unlike tooltips, which may behave differently based on the device or browser, toggle sections offer greater consistency in functionality and interaction.

**Contextual Relevance**: Toggle sections can be strategically placed within the interface to provide relevant information in context. Users can expand the section to access additional details or explanations related to the content they are viewing, enhancing their understanding and engagement.

**Accessibility**: Toggle sections can be designed to be accessible to users with disabilities, such as screen readers or keyboard navigation, ensuring equal access to information for all users.

While toggle sections offer several advantages as an alternative to tooltips, it's essential to consider the specific needs of the application or website and the preferences of the target audience. Depending on the context and content requirements, tooltips, toggle sections, or a combination of both may be the most suitable choice for enhancing the user experience on mobile or touch devices.

----
```
context: tooltips on popups in mobile devices
Now which alternative is most suitable
```

Given the context of tooltips on popups in mobile devices, the most suitable alternative depends on various factors such as the complexity of the content, user preferences, and the design constraints of the application or website. Here are some alternative approaches that may be more appropriate:

**Inline Help Text within Popups**: Instead of using tooltips, consider incorporating brief descriptions or explanations directly within the popups. This approach ensures that users can access relevant information without relying on hover interactions, making it more compatible with mobile devices.

**Progressive Disclosure within Popups**: Implement progressive disclosure techniques within popups to gradually reveal information or options as users interact with them. Start with essential content or actions and provide options to access more detailed information or advanced features as needed.

**Embedded Videos or Demos within Popups**: Include embedded videos or interactive demos within popups to demonstrate how to use specific features or accomplish tasks. Videos and demos provide visual guidance and can be particularly effective in engaging users within the context of popups.

**Contextual Menus within Popups**: Integrate contextual menus within popups to offer additional options or actions relevant to the content or functionality presented. Contextual menus allow users to access relevant commands quickly without cluttering the popup interface.

User Feedback and Support Options within Popups: Provide options for users to access help, support, or feedback mechanisms directly within the popups. This ensures that users can seek assistance or provide feedback without navigating away from the context of the popup.

**Reply by benoitgrelard** [reply, marked as answer]
Yeah I suspected it would be something like that. That's definitely something we don't recommend, typically a tooltip is placed over a "tool" which is typically a button, so putting the tooltip on click doesn't make sense as that's generally what triggers the action too (negating the point of having a tooltip). In these cases we generally recommend a popover with a little icon button as a trigger.
